training/polynomial_features.py

```diff
@@ -0,0 +1,208 @@
+
+
+def sample_keys_train_val(
+    dict_dataset: DictDataset, percentage_val=0.2
+) -> Tuple[DictDataset, DictDataset, np.ndarray]:
+    """Returns data split by train and validation trajectories.
+
+    Parameters
+    ----------
+    dict_dataset : DictDataset
+        Dataset with all trajectories.
+    percentage_val : float, optional
+        Percentage of trajectories to use in validation, by default 0.2
+
+    Returns
+    -------
+    Tuple[DictDataset, DictDataset, np.ndarray]
+        Train and validation dataset and keys of trajectories defined for validation.
+    """
+    list_of_keys = list(dict_dataset.keys())
+    validation_keys = np.random.choice(
+        list_of_keys, size=int(len(dict_dataset) * percentage_val), replace=False
+    )
+    train_dict = {}
+    validation_dict = {}
+
+    for key in list_of_keys:
+        if key in validation_keys:
+            validation_dict[key] = dict_dataset[key]
+        else:
+            train_dict[key] = dict_dataset[key]
+    return train_dict, validation_dict, validation_keys
+
+
+def convert_dict_to_np_dataset(
+    dict_dataset: DictDataset, include_delta=False
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Converts dictionary of datasets into numpy arrays used for training the model.
+
+    Parameters
+    ----------
+    dict_dataset : DictDataset
+        Datasets organized by trajectory
+    include_delta : bool, optional
+        Whether to include the delta of positions between steps, by default False
+
+    Returns
+    -------
+    Tuple[np.ndarray, np.ndarray]
+        Trajectory data joined in a numpy array.
+    """
+    X = []
+    y = []
+    for _, traj_df in dict_dataset.items():
+        if include_delta:
+            X.append(
+                np.concatenate(
+                    [
+                        traj_df.iloc[:-1][state_columns].to_numpy(),
+                        traj_df.iloc[1:][next_positions].to_numpy(),
+                        traj_df.iloc[1:][next_positions].to_numpy()
+                        - traj_df.iloc[:-1][next_positions].to_numpy(),
+                        traj_df.iloc[:-1][control_input_columns].to_numpy(),
+                    ],
+                    axis=1,
+                )
+            )
+        else:
+            X.append(
+                np.concatenate(
+                    [
+                        traj_df.iloc[:-1][state_columns].to_numpy(),
+                        traj_df.iloc[1:][next_positions].to_numpy(),
+                        traj_df.iloc[:-1][control_input_columns].to_numpy(),
+                    ],
+                    axis=1,
+                )
+            )
+        y.append(traj_df.iloc[1:][state_columns].to_numpy())
+
+    X_arr = np.concatenate(X, axis=0)
+    y_arr = np.concatenate(y, axis=0)
+    return X_arr, y_arr
+
+
+if __name__ == "__main__":
+    systems_path = "./training_trajectories/systems"
+
+    with open(systems_path, "r") as f:
+        systems_list = [i.strip() for i in f.readlines()]
+
```

Comment:
I believe here we should exclude the bumblebee systems, as those are trained with the other code.

Reply:
Good point. I excluded beetle and butterfly systems from the version of the training I committed to this repository.